library/std/src/os/unix/net/ancillary/ip.rs

```diff
@@ -0,0 +1,330 @@
+            self.send_vectored_with_ancillary_to(
+                msg_name as *mut libc::c_void,
+                msg_namelen,
+                bufs,
+                &mut ancillary.inner,
+            )
+        }
+    }
+}
+
+/// This enum represent one control message of variable type for `IPPROTO_IP`.
+#[unstable(feature = "unix_socket_ancillary_data", issue = "76915")]
+#[non_exhaustive]
+pub enum IpAncillaryData {
+    Ttl(u8),
+    HopLimit(u8),
+}
+
+impl IpAncillaryData {
+    /// Convert the a `libc::c_int` to `u8`
+    ///
+    /// # Safety
+    ///
+    /// `data` must contain at least one `libc::c_int`.
+    unsafe fn as_u8(data: &[u8]) -> u8 {
+        let mut ancillary_data_iter = AncillaryDataIter::<libc::c_int>::new(data);
+        if let Some(u) = ancillary_data_iter.next() { u as u8 } else { 0 }
+    }
+
+    /// Create a `AncillaryData::Ttl` variant.
+    ///
+    /// # Safety
+    ///
+    /// `data` must contain a valid control message and the control message must be type of
+    /// `IPPROTO_IP` and level of `IP_TTL`.
+    unsafe fn as_ttl(data: &[u8]) -> Self {
+        let ttl = IpAncillaryData::as_u8(data);
+        IpAncillaryData::Ttl(ttl)
+    }
+
+    /// Create a `AncillaryData::HopLimit` variant.
+    ///
+    /// # Safety
+    ///
+    /// `data` must contain a valid control message and the control message must be type of
+    /// `IPPROTO_IPV6` and level of `IPV6_HOPLIMIT`.
+    unsafe fn as_hop_limit(data: &[u8]) -> Self {
+        let hop_limit = IpAncillaryData::as_u8(data);
+        IpAncillaryData::HopLimit(hop_limit)
+    }
+
+    unsafe fn try_from(cmsg: &libc::cmsghdr) -> Result<Self, AncillaryError> {
+        let data = get_data_from_cmsghdr(cmsg);
+
+        match (*cmsg).cmsg_level {
+            libc::IPPROTO_IP => match (*cmsg).cmsg_type {
+                libc::IP_TTL => Ok(IpAncillaryData::as_ttl(data)),
+                cmsg_type => {
+                    Err(AncillaryError::Unknown { cmsg_level: libc::IPPROTO_IP, cmsg_type })
+                }
+            },
+            libc::IPPROTO_IPV6 => match (*cmsg).cmsg_type {
+                libc::IPV6_HOPLIMIT => Ok(IpAncillaryData::as_hop_limit(data)),
+                cmsg_type => {
+                    Err(AncillaryError::Unknown { cmsg_level: libc::IPPROTO_IPV6, cmsg_type })
+                }
+            },
+            cmsg_level => Err(AncillaryError::Unknown { cmsg_level, cmsg_type: (*cmsg).cmsg_type }),
+        }
+    }
+}
+
+#[unstable(feature = "unix_socket_ancillary_data", issue = "76915")]
+impl<'a> Iterator for Messages<'a, IpAncillaryData> {
+    type Item = Result<IpAncillaryData, AncillaryError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        unsafe {
+            let cmsg = self.next_cmsghdr()?;
+            let ancillary_result = IpAncillaryData::try_from(cmsg);
+            Some(ancillary_result)
+        }
+    }
+}
+
+/// A net IP Ancillary data struct.
```

Comment:
This needs several paragraphs of text and links to external documentation (e.g. RFC 3542, cmsg manpages or posix specs)